File: clients/android/app/src/main/java/com/syncbridge/android/network/NetworkManager.kt
package com.syncbridge.android.network

import android.content.Context
import android.content.SharedPreferences
import com.syncbridge.android.data.ApiClient
import com.syncbridge.android.data.DeviceEntry
import com.syncbridge.android.data.DiagnosticsResponse
import com.syncbridge.android.data.LocalPeer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

data class SignalPeerEvent(
    val deviceId: String,
    val addrs: List<String>,
    val port: Int,
    val at: String,
)

data class EnrichedPeer(
    val deviceId: String,
    val addrs: List<String>,
    val port: Int,
    val updatedAt: String,
    val name: String,
    val platform: String,
    val connectionType: TransferRoute,
)

data class TransferLogEntry(
    val id: String,
    val at: String,
    val name: String,
    val method: TransferRoute,
    val fallbackReason: String? = null,
    val bytesPerSec: Long? = null,
    val peerDeviceId: String? = null,
)

data class NetworkSnapshot(
    val diagnostics: DiagnosticsResponse? = null,
    val wsConnected: Boolean = false,
    val peers: List<LocalPeer> = emptyList(),
    val devices: List<DeviceEntry> = emptyList(),
    val enrichedPeers: List<EnrichedPeer> = emptyList(),
    val lastSignalTime: String? = null,
    val nearbyAlert: SignalPeerEvent? = null,
    val currentTransferMode: String = "Automatic",
    val lastSyncAt: String? = null,
    val latencyMs: Long? = null,
    val transferLogs: List<TransferLogEntry> = emptyList(),
    val loading: Boolean = false,
    val error: String? = null,
)

class NetworkManager(
    private val context: Context,
    private val api: ApiClient,
    private val prefs: SharedPreferences,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val _state = MutableStateFlow(NetworkSnapshot())
    val state: StateFlow<NetworkSnapshot> = _state.asStateFlow()

    private var running = false
    private var clientIp = ""
    private val knownPeerIds = mutableSetOf<String>()

    fun start() {
        if (running) return
        running = true
        scope.launch { refreshLoop() }
        scope.launch { advertiseLoop() }
    }

    fun stop() {
        running = false
        knownPeerIds.clear()
        _state.value = NetworkSnapshot()
    }

    fun setWsConnected(connected: Boolean) {
        _state.update {
            it.copy(
                wsConnected = connected,
                currentTransferMode = NetworkPreferences.systemTransferLabel(connected),
            )
        }
    }

    fun markSync() {
        _state.update { it.copy(lastSyncAt = java.time.Instant.now().toString()) }
    }

    fun dismissNearbyAlert() {
        _state.update { it.copy(nearbyAlert = null) }
    }

    suspend fun refreshOncePublic() {
        refreshOnce()
    }

    /** @return true the first time we see this peer (for optional toast). */
    fun handleSignalPeer(deviceId: String, addrs: List<String>, port: Int): Boolean {
        val isNew = !knownPeerIds.contains(deviceId)
        if (isNew) knownPeerIds.add(deviceId)
        val event = SignalPeerEvent(
            deviceId = deviceId,
            addrs = addrs,
            port = port,
            at = java.time.Instant.now().toString(),
        )
        _state.update {
            it.copy(
                lastSignalTime = event.at,
                nearbyAlert = if (isNew) event else it.nearbyAlert,
            )
        }
        return isNew
    }

    fun resolveUploadRoute(
        fileSizeBytes: Long,
        fileCount: Int = 1,
        isFolder: Boolean = false,
    ): UploadRoute {
        val r = FileRouting.resolve(fileSizeBytes, fileCount, isFolder)
        return UploadRoute(r.transferMode, r.route, r.fallbackReason)
    }

    fun logTransfer(
        name: String,
        method: TransferRoute,
        fallbackReason: String? = null,
        bytesPerSec: Long? = null,
        peerDeviceId: String? = null,
    ) {
        val entry = TransferLogEntry(
            id = java.util.UUID.randomUUID().toString(),
            at = java.time.Instant.now().toString(),
            name = name,
            method = method,
            fallbackReason = fallbackReason,
            bytesPerSec = bytesPerSec,
            peerDeviceId = peerDeviceId,
        )
        _state.update { it.copy(transferLogs = (listOf(entry) + it.transferLogs).take(50)) }
    }

    private suspend fun refreshLoop() {
        while (scope.isActive && running) {
            refreshOnce()
            delay(15_000)
        }
    }

    private suspend fun advertiseLoop() {
        while (scope.isActive && running) {
            delay(60_000)
            if (!running) return
            advertiseOnce()
        }
    }

    private suspend fun refreshOnce() {
        _state.update { it.copy(loading = true, error = null) }
        val t0 = System.currentTimeMillis()
        try {
            val diagnostics = api.fetchDiagnostics()
            clientIp = diagnostics.clientIp
            advertiseOnce()
            val localAddrs = LanAddressHelper.localIpv4Addresses()
            val addrsQuery = (localAddrs + listOfNotNull(diagnostics.clientIp.takeIf { it.isNotBlank() }))
                .distinct().joinToString(",")
            val peerData = api.fetchLocalPeers(addrsQuery)
            val deviceData = runCatching { api.fetchDevices() }.getOrDefault(emptyList())
            val peers = peerData
            val enriched = enrichPeers(peers, deviceData)
            val latencyMs = System.currentTimeMillis() - t0
            _state.update {
                it.copy(
                    diagnostics = diagnostics,
                    peers = peers,
                    devices = deviceData,
                    enrichedPeers = enriched,
                    latencyMs = latencyMs,
                    loading = false,
                    currentTransferMode = NetworkPreferences.systemTransferLabel(it.wsConnected),
                )
            }
        } catch (e: Exception) {
            _state.update {
                it.copy(loading = false, error = e.message ?: "Network refresh failed")
            }
        }
    }

    private suspend fun advertiseOnce() {
        val addrs = LanAddressHelper.localIpv4Addresses()
        if (addrs.isNotEmpty()) {
            runCatching { api.advertiseLocalAddrs(addrs) }
        }
    }

    private fun enrichPeers(peers: List<LocalPeer>, devices: List<DeviceEntry>): List<EnrichedPeer> {
        val byId = devices.associateBy { it.id }
        return peers.map { p ->
            val dev = byId[p.deviceId]
            EnrichedPeer(
                deviceId = p.deviceId,
                addrs = p.addrs,
                port = p.port,
                updatedAt = p.updatedAt,
                name = dev?.name ?: "${p.deviceId.take(8)}…",
                platform = dev?.platform ?: "unknown",
                connectionType = TransferRoute.DirectLan,
            )
        }
    }

    data class UploadRoute(
        val transferMode: String,
        val route: TransferRoute,
        val fallbackReason: String? = null,
        val peerDeviceId: String? = null,
    )

    companion object {
        fun platformLabel(platform: String): String = when (platform) {
            "macos" -> "Mac"
            "android" -> "Android"
            "ios" -> "iPhone"
            "web" -> "Web"
            else -> platform
        }
    }
}

File: clients/macos/SyncBridgeMac/Core/Network/NetworkManager.swift
// NetworkManager.swift — LAN discovery, routing, diagnostics logging

import Foundation
import Network
import Darwin

struct SignalPeerEvent: Equatable {
    let deviceId: String
    let addrs: [String]
    let port: Int
    let at: String
}

struct TransferLogEntry: Identifiable {
    let id: String
    let at: String
    let name: String
    let method: String
    let fallbackReason: String?
    let bytesPerSec: Int64?
    let peerDeviceId: String?
}

struct EnrichedPeer: Identifiable {
    let deviceId: String
    let addrs: [String]
    let port: Int
    let updatedAt: String
    let name: String
    let platform: String
    var id: String { deviceId }
}

@MainActor
final class NetworkManager: ObservableObject {

    @Published private(set) var diagnostics: DiagnosticsResponse?
    @Published private(set) var peers: [LocalPeerResponse] = []
    @Published private(set) var enrichedPeers: [EnrichedPeer] = []
    @Published private(set) var lastSignalTime: String?
    @Published private(set) var nearbyAlert: SignalPeerEvent?
    @Published private(set) var currentTransferMode = "Cloud Relay"
    @Published private(set) var lastSyncAt: String?
    @Published private(set) var latencyMs: Int?
    @Published private(set) var transferLogs: [TransferLogEntry] = []
    @Published private(set) var loading = false
    @Published private(set) var errorMessage: String?

    private var knownPeerIds = Set<String>()

    var wsConnected = false {
        didSet { updateTransferLabel() }
    }

    private let auth: AuthService
    private var refreshTask: Task<Void, Never>?
    private var advertiseTask: Task<Void, Never>?
    private var clientIp = ""
    private var uiActive = true

    init(auth: AuthService) {
        self.auth = auth
    }

    /// Slow or skip LAN polling when the app is not active (saves CPU/battery).
    func setUiActive(_ active: Bool) {
        uiActive = active
        if active {
            Task { await refresh() }
        }
    }

    func start() {
        stop()
        refreshTask = Task { [weak self] in
            while !Task.isCancelled {
                if self?.uiActive == true {
                    await self?.refresh()
                }
                let delayNs: UInt64 = self?.uiActive == true ? 15_000_000_000 : 300_000_000_000
                try? await Task.sleep(nanoseconds: delayNs)
            }
        }
        advertiseTask = Task { [weak self] in
            while !Task.isCancelled {
                let delayNs: UInt64 = self?.uiActive == true ? 60_000_000_000 : 300_000_000_000
                try? await Task.sleep(nanoseconds: delayNs)
                if self?.uiActive == true {
                    await self?.advertiseOnce()
                }
            }
        }
    }

    func stop() {
        refreshTask?.cancel()
        advertiseTask?.cancel()
        refreshTask = nil
        advertiseTask = nil
        diagnostics = nil
        peers = []
        enrichedPeers = []
        nearbyAlert = nil
        knownPeerIds.removeAll()
        wsConnected = false
    }

    func markSync() {
        lastSyncAt = ISO8601DateFormatter().string(from: Date())
    }

    func dismissNearbyAlert() {
        nearbyAlert = nil
    }

    /// Returns true the first time we see this peer (for optional notification).
    @discardableResult
    func handleSignalPeer(deviceId: String, addrs: [String], port: Int) -> Bool {
        let isNew = !knownPeerIds.contains(deviceId)
        if isNew { knownPeerIds.insert(deviceId) }
        let event = SignalPeerEvent(
            deviceId: deviceId,
            addrs: addrs,
            port: port,
            at: ISO8601DateFormatter().string(from: Date())
        )
        lastSignalTime = event.at
        if isNew { nearbyAlert = event }
        return isNew
    }

    func resolveUploadRoute(
        fileSizeBytes: Int,
        fileCount: Int = 1,
        isFolder: Bool = false
    ) -> (transferMode: String, route: String, fallback: String?, peerId: String?) {
        let r = FileRouting.resolve(fileSizeBytes: fileSizeBytes, fileCount: fileCount, isFolder: isFolder)
        return (r.transferMode, r.route, r.fallback, nil)
    }

    func logTransfer(name: String, method: String, fallback: String?, bytesPerSec: Int64?, peerId: String?) {
        let entry = TransferLogEntry(
            id: UUID().uuidString,
            at: ISO8601DateFormatter().string(from: Date()),
            name: name,
            method: method,
            fallbackReason: fallback,
            bytesPerSec: bytesPerSec,
            peerDeviceId: peerId
        )
        transferLogs = ([entry] + transferLogs).prefix(50).map { $0 }
    }

    func refresh() async {
        loading = true
        errorMessage = nil
        let t0 = CFAbsoluteTimeGetCurrent()
        do {
            let diag = try await auth.fetchDiagnostics()
            diagnostics = diag
            clientIp = diag.clientIp
            await advertiseOnce()
            let localAddrs = Self.localIPv4Addresses()
            let query = (localAddrs + [diag.clientIp].filter { !$0.isEmpty }).joined(separator: ",")
            let peerResp = try await auth.fetchLocalPeers(addrs: query)
            peers = peerResp.peers
            let devices = try await auth.listDevices().devices
            enrichedPeers = enrich(peers: peerResp.peers, devices: devices)
            latencyMs = Int((CFAbsoluteTimeGetCurrent() - t0) * 1000)
            loading = false
            updateTransferLabel()
        } catch {
            loading = false
            errorMessage = error.localizedDescription
        }
    }

    private func advertiseOnce() async {
        let addrs = Self.localIPv4Addresses()
        guard !addrs.isEmpty else { return }
        _ = try? await auth.advertiseLocalAddrs(addrs)
    }

    private func updateTransferLabel() {
        currentTransferMode = wsConnected ? "Automatic" : "Offline"
    }

    private func enrich(peers: [LocalPeerResponse], devices: [DeviceResponse]) -> [EnrichedPeer] {
        let byId = Dictionary(uniqueKeysWithValues: devices.map { ($0.id, $0) })
        return peers.map { p in
            let dev = byId[p.deviceId]
            return EnrichedPeer(
                deviceId: p.deviceId,
                addrs: p.addrs,
                port: p.port,
                updatedAt: p.updatedAt,
                name: dev?.name ?? String(p.deviceId.prefix(8)) + "…",
                platform: dev?.platform ?? "unknown"
            )
        }
    }

    static func localIPv4Addresses() -> [String] {
        var addrs: [String] = []
        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddr) == 0, let first = ifaddr else { return [] }
        defer { freeifaddrs(ifaddr) }
        for ptr in sequence(first: first, next: { $0.pointee.ifa_next }) {
            let iface = ptr.pointee
            guard iface.ifa_addr.pointee.sa_family == UInt8(AF_INET) else { continue }
            let name = String(cString: iface.ifa_name)
            if name == "lo0" { continue }
            var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
            getnameinfo(iface.ifa_addr, socklen_t(iface.ifa_addr.pointee.sa_len),
                        &hostname, socklen_t(hostname.count), nil, 0, NI_NUMERICHOST)
            let ip = String(cString: hostname)
            if !ip.isEmpty { addrs.append(ip) }
        }
        return addrs
    }

    static func platformLabel(_ platform: String) -> String {
        switch platform {
        case "macos": return "Mac"
        case "android": return "Android"
        case "ios": return "iPhone"
        case "web": return "Web"
        default: return platform
        }
    }
}
